### ISSUE TITLE
If the input contains ^^ it only shows the first ^, the second is still in the input buffer but input only progresses on next letter

### ISSUE THREAD

**Issue by kasterma** [Miodec/monkeytype#2536]
### Did you clear cache before opening an issue?

- [X] I have cleared my cache

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Does the issue happen when logged in?

Yes

### Does the issue happen when logged out?

Yes

### Does the issue happen in incognito mode when logged in?

Yes

### Does the issue happen in incognito mode when logged out?

Yes

### Current Behavior

I find that in all my test setups entering ^^ in a row only progresses the cursor one place until the next character is typed.  Then all the "missing" ^ are used to progress the cursor, but not to color the letters progressed over.

### Expected Behavior

On every letter typed I would expect the cursor to progress one letter.

### Steps To Reproduce

1. start any test
2. type multiple ^
3. now you see the cursor not progress as expected
4. type any other letter
5. now you see the cursor progress over the right number of letters, but not color them

### Environment

- OS: OS X 12.2.1
- Browser: firefox 97.0.1, safari 15.3

**Comment by omackenzie** [comment]
This seems to be a problem with the "trailingComposeChars" function in misc. It detects the ^ character and then doesn't update the position of the caret in input-controller. The same problem occurs for the ` and ~ characters.

**Comment by SeerLite** [comment]
That's intended, at least in the way it currently works. It's used to not skip ahead when entering dead characters in input systems that show a temporary character in-place.

I realized some time after that there's an easier way to do it by using an event/property built in to JS to check if the user is currently in "insertion mode" or whatever it's called.

I'll try to send a PR to use this new system soon. I'm currently trying to update my offline fork with the new repo structure so I can actually do anything. I'll get there, but give me some time.

@Miodec assign to me!

**Comment by Miodec** [comment]
@SeerLite just looking into this - do you mean these properties on the InputEvent?

![image](https://user-images.githubusercontent.com/13181393/168445450-2b1fa47b-ec36-48b3-9b55-b6ec252c835a.png)

**Comment by SeerLite** [comment]
Hi Mio sorry for not answering again. Yes, those should be the properties. I don't remember exactly if you should go for the properties _or_ the events (there's a "composing started" event). One of the two isn't consistent across browsers so you have to pick the other one. IIRC I wrote about this in the dev channel on Discord last year, maybe you can find it.

And sorry for overpromising again, didn't think I'd be _this_ busy with uni. So please unassign me, I don't think I'll be able to contribute to the project again anytime soon.

Also, I'm very happy to see the progress you've been making :) I've been noticing many "monkeytype-inspired" projects lately and it made me realize just how big MT has gotten. I wish the best for you and the site. <3

**Comment by omackenzie** [comment]
I'm happy to work on this, if no one else wants to.

**Comment by Miodec** [comment]
@SeerLite alright mate, good luck and stuff (also, I cant find your message on Discord :|)

@omackenzie alright!

**Comment by Miodec** [comment]
From Discord:
```
so looking for a solution i found out that this is a thing apparently?? https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event
oh god does this catch absolutely everything
```
![image](https://user-images.githubusercontent.com/13181393/169651748-db8b6594-02cc-4f21-b199-f963f70dff99.png)
```
if we figure this out we could use it for all the composed languages
```
![image](https://user-images.githubusercontent.com/13181393/169651752-bc609035-0b13-4797-b390-27ab6b2479ff.png)
```
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/isComposing this one seems to be more accurate
oh god all my hard work on trailingcomposechars
and there was a builtin thing to check it
only if we use the events instead of this ^
or if you mean the safari events then yeah
we would have to do the workaround of delaying the input event to force keydown to happen before
still, it looks like it's gonna be a lot more consistent if we use this property instead of regex 
```

**Comment by Miodec** [comment]
The main thing right now is check consistency between platforms:
 - iOS
 - Android
 - Windows
 - Mac
 - Linux

**Comment by omackenzie** [comment]
As part of the new system, I can make it so that all composing characters appear the same as dead keys when incomplete, rather than showing it as an error. Would you say this looks better?

New:
https://user-images.githubusercontent.com/30273552/170244996-73360bfb-5847-440c-8e6e-65377f3ab54b.mp4

Old:
https://user-images.githubusercontent.com/30273552/170245266-2a401869-3b2c-43c8-8145-48d0fca40996.mp4

Note that this is purely visual, it doesn't fix the accuracy problem. I've tried a method of ignoring the accuracy calculations while composing, then checking all the input together once the composition ends. This is promising for languages like Korean, where the composition ends after each individual block, but this doesn't work as well with Japanese. You can essentially type a whole sentence while composing and the accuracy isn't checked until it is entered. So you could make a mistake, go back and fix it, and it wouldn't be counted if you never clicked enter.
So yeah, could possibly be useful. I won't be implementing it for now though, just an idea!

**Comment by Miodec** [comment]
Yeah thats good. By the way if you want to discuss this in real time, feel free to join the Discord server and message in the #development channel (discord.gg/monkeytype)
